### PR TITLE
Extract out simplification changes

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Pulling out independent parts of https://github.com/dotnet/roslyn/pull/40746 to make that PR simpler.

This PR should be reviewed with whitespace off.

Specific changes:

1. Unify patterns in extension methods.
2. Convert private extensions called from 1 location into normal static methods.
3. Make public extensions only called from one place private.
4. Compute symbol being reduced only once and pass along to all simplification helpers.
5. Remove codepaths dealing with 'candidate' symbols.  Later code bails out with ambiguities, so no need to deal with them earlier.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @sharwell

File: src/Workspaces/Core/Portable/Simplification/SimplificationHelpers.cs [hunk 1]
```diff
@@ -101,15 +101,16 @@ internal static bool IsValidSymbolInfo(ISymbol symbol)
...
         {
-            var symbol = GetOriginalSymbolInfo(semanticModel, expression);
-            if (symbol == null ||
-                (!symbol.IsStatic &&
-                 (symbol.IsKind(SymbolKind.Field) && optionSet.GetOption(CodeStyleOptions.QualifyFieldAccess, semanticModel.Language).Value ||
-                 (symbol.IsKind(SymbolKind.Property) && optionSet.GetOption(CodeStyleOptions.QualifyPropertyAccess, semanticModel.Language).Value) ||
-                 (symbol.IsKind(SymbolKind.Method) && optionSet.GetOption(CodeStyleOptions.QualifyMethodAccess, semanticModel.Language).Value) ||
-                 (symbol.IsKind(SymbolKind.Event) && optionSet.GetOption(CodeStyleOptions.QualifyEventAccess, semanticModel.Language).Value))))
+            if (symbol.IsStatic)
+                return false;
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Simplification/SimplificationHelpers.cs]: ❔  In the old code, this case returned `true`. Was the change intentional?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Simplification/SimplificationHelpers.cs]: Fixed.  Thanks for catching this.  Is also revealed a test hold.  I've reverted to the proper logic and added a test to ensure that we don't regrss this desirable behavior.

**Comment by CyrusNajmabadi** [comment]
@sharwell @dpoeschl this is ready for review.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 2]
```diff
@@ -918,28 +920,12 @@ public static SimpleNameSyntax GetNameWithTriviaMoved(this MemberAccessExpressio
...
-            bool containsOtherSymbol;
-            if (actualSymbol.Symbol is object)
-            {
-                containsNamespaceOrTypeSymbol = actualSymbol.Symbol is INamespaceOrTypeSymbol;
-                containsOtherSymbol = !containsNamespaceOrTypeSymbol;
-            }
-            else if (!actualSymbol.CandidateSymbols.IsDefaultOrEmpty)
-            {
-                containsNamespaceOrTypeSymbol = actualSymbol.CandidateSymbols.Any(symbol => symbol is INamespaceOrTypeSymbol);
-                containsOtherSymbol = actualSymbol.CandidateSymbols.Any(symbol => !(symbol is INamespaceOrTypeSymbol));
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: 💭 I'm a bit concerned about this change. Is it observable?

File: src/Workspaces/Core/Portable/Simplification/SimplificationHelpers.cs [hunk 3]
```diff
@@ -102,15 +102,15 @@ internal static bool IsValidSymbolInfo(ISymbol symbol)
...
+
+            if ((symbol.IsKind(SymbolKind.Field) && optionSet.GetOption(CodeStyleOptions.QualifyFieldAccess, semanticModel.Language).Value ||
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Simplification/SimplificationHelpers.cs]: 💭 It's strange to see the outermost parenthesized expression going around the entire condition, as opposed to just surrounding the first `&&` expression.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 4]
```diff
@@ -815,8 +815,12 @@ private static bool CanReplace(ISymbol symbol)
                 return true;
             }
 
+            var symbol = SimplificationHelpers.GetOriginalSymbolInfo(semanticModel, memberAccess);
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: 📝 There are two potential behavior changes here. They occur if:

1. `memberAccess` does not have `SpecialTypeAnnotation` but `memberAccess.Name` _does_ have `SpecialTypeAnnotation`. In this case, the `SpecialTypeAnnotation` used to be considered by `ShouldSimplifyMemberAccessExpression` but is now ignored.
2. `memberAccess.Name` has `SymbolAnnotation`, but `memberAccess` either does not have the annotation, or has the annotation and it points to a different symbol. In this case, `ShouldSimplifyMemberAccessExpression` will consider the annotation on `memberAccess` where before the change it considered the annotation on `memberAccess.Name`.

File: src/Workspaces/Core/Portable/Simplification/SimplificationHelpers.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Simplification/SimplificationHelpers.cs]: ick yes.  nice catch.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs (hunk 2 above)
Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: Here's a case where it _might_ go on this code path:

```
class Example
{
    void Method()
    {
        _ = nameof(this.ToString); // Should offer to simplify to `nameof(ToString)`
    }

    public override string ToString() => throw null;
    public string ToString(string format, IFormatProvider formatProvider) => throw null;
}
```

Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: ➡️ This issue was resolved by filing follow-up issues. The current code handled ambiguous symbols, but other parts of the simplification system did not and in the end it did not work prior to this change.

#40972
#40973

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs (hunk 4 above)
Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: I reviewed the usage of `SpecialTypeAnnotation` and `SymbolAnnotation`, and found that neither of them get applied to the `Name` node of a `this.Name` or `Me.Name` expression.

**Comment by CyrusNajmabadi** [comment]
@sharwell feel free to merge.

**Pull request merged**